### PR TITLE
Report startup errors in the Gateway StatusFailure

### PR DESCRIPTION
**Pull request by tpantelis**
Issue #777 

Added a `SetGatewayStatusError` method to the `Gateway` syncer that creates/updates the `Gateway` resource with a specific `StatusFailure`. This is called prior to exiting via `Fatalf`.

This doesn't include some bootstrap failures that prevent the syncer from being created, like failure to create the kube client.

Eliminated a couple error paths that never actually result in an error to simplify:

 - `util.GetLocalCluster` - this function simply copies fields so I eliminated it altogether and inlined it.

 - `cableengine.NewEngine` - removed the error return - any errors should be reported by `StartEngine`..

### CODE REVIEW

**Review by skitt — APPROVED** [review]

File: main.go [hunk 1]
```diff
@@ -92,9 +94,15 @@ func main() {
...
-		klog.Fatalf("Error creating local cluster object from %#v: %v", submSpec, err)
+	localCluster := types.SubmarinerCluster{
+		ID: submSpec.ClusterID,
+		Spec: subv1.ClusterSpec{
+			ClusterID:   submSpec.ClusterID,
+			ColorCodes:  submSpec.ColorCodes,
+			ServiceCIDR: submSpec.ServiceCidr,
+			ClusterCIDR: submSpec.ClusterCidr,
+			GlobalCIDR:  submSpec.GlobalCidr,
+		},
```
Comment (mangelajo) [inline main.go:105]: Just for having a shorter main, I'd keep it on a function, may be `localCluster := GetClusterFromSpec(submSpec)`

of course, without return of error which made no sense.

File: main.go [hunk 2]
```diff
@@ -268,4 +277,12 @@ func startLeaderElection(leaderElectionClient kubernetes.Interface, recorder res
 			OnStoppedLeading: end,
 		},
 	})
+
+	return nil
+}
+
+func fatal(gwSyncer *syncer.GatewaySyncer, format string, args ...interface{}) {
```
Comment (mangelajo) [inline main.go:284]: I think this approach is great :+1:

File: pkg/cableengine/cableengine.go [hunk 3]
```diff
@@ -48,14 +48,12 @@ type engine struct {
...
-func NewEngine(localCluster types.SubmarinerCluster, localEndpoint types.SubmarinerEndpoint) (Engine, error) {
-	i := engine{
+func NewEngine(localCluster types.SubmarinerCluster, localEndpoint types.SubmarinerEndpoint) Engine {
+	return &engine{
 		localCluster:  localCluster,
 		localEndpoint: localEndpoint,
 		driver:        nil,
 	}
-
-	return &i, nil
```
Comment (mangelajo) [inline pkg/cableengine/cableengine.go:58]: why did we end up with this? very weird...

File: pkg/cableengine/syncer/syncer.go [hunk 4]
```diff
@@ -49,6 +52,21 @@ func (s *GatewaySyncer) Run(stopCh <-chan struct{}) {
 }
 
 func (i *GatewaySyncer) syncGatewayStatus() {
+	i.Lock()
+	defer i.Unlock()
+
+	i.syncGatewayStatusSafe()
```
Comment (mangelajo) [inline pkg/cableengine/syncer/syncer.go:58]: :+1:

**Review by mangelajo — CHANGES_REQUESTED** [review]
Just a small nit, but the change in general makes it 10 times more robust.

File: main.go (hunk 1 above)
Comment (tpantelis) [reply main.go]: Done

File: main.go [hunk 5]
```diff
@@ -92,10 +94,7 @@ func main() {
 
 	klog.Info("Creating the cable engine")
 
-	localCluster, err := util.GetLocalCluster(submSpec)
-	if err != nil {
-		klog.Fatalf("Error creating local cluster object from %#v: %v", submSpec, err)
-	}
+	localCluster := submarinerClusteFrom(&submSpec)
```
Comment (skitt) [inline main.go:97]:
```suggestion
	localCluster := submarinerClusterFrom(&submSpec)
```

File: main.go [hunk 6]
```diff
@@ -198,18 +193,36 @@ func main() {
...
+		if err = startLeaderElection(leClient, recorder, becameLeader, lostLeader); err != nil {
+			fatal(cableEngineSyncer, "Error starting leader election: %v", err)
+		}
+	}()
+
 	<-stopCh
 	klog.Info("All controllers stopped or exited. Stopping main loop")
 }
 
+func submarinerClusteFrom(submSpec *types.SubmarinerSpecification) types.SubmarinerCluster {
```
Comment (skitt) [inline main.go:206]:
```suggestion
func submarinerClusterFrom(submSpec *types.SubmarinerSpecification) types.SubmarinerCluster {
```

**Review by skitt — CHANGES_REQUESTED** [review]
Typo ;-).

File: main.go (hunk 6 above)
Comment (tpantelis) [reply main.go:206]: > Typo ;-).

I had just woke up and didn't have my readers on 😄

**Review by skitt — APPROVED** [review]

**Review by mangelajo — APPROVED** [review]

**Pull request merged**